### DISCUSSION TITLE
How to use Json files as environment variables with vercel

### DISCUSSION THREAD

**Discussion by TravisC671** [vercel/vercel#6782] [Help] [answered]
Hello! I have just set up Vercel and one of my environment variables on my local directory is `GOOGLE_APPLICATION_CREDENTIALS=./secrets.json`
and that links to secrets.json which contains all my oath and API keys. How would I set that up with Vercel in a secure way? I have tried just pasting the JSON file but it does not like that and gives a huge error.

**Comment by leerob** [comment, marked as answer]
You could encrypt the JSON file and add it as an Environment Variable: https://vercel.com/support/articles/how-do-i-workaround-vercel-s-4-kb-environment-variables-limit

**Reply by TravisC671** [reply]
Hello, I have been looking at this, and it seems like the correct answer, but is there a way to set the unencrypted variable to an environment variable?
https://cloud.google.com/docs/authentication/getting-started#windows
for the Google API, it looks explicitly for the GOOGLE_APPLICATION_CREDENTIALS environment variable.  What would be the right way to approach this?

**Reply by TravisC671** [reply]
Or maybe I could decrypt and edit the JSON that GOOGLE_APPLICATION_CREDENTIALS points to at build time. That way, my IV and my key would be left out of GitHub but kept in the final build. Is that possible?

**Reply by hayata-suenaga** [reply]
I cannot access the link. I'd appreciate it if someone could point me to the right resource 🙇

**Comment by arlbibek** [comment]
Hi @TravisC671, did you find a solution (or a workaround)? If you did can you please share it here?

**Comment by zachgoll** [comment]
Just wanted to share a solution that I came up with that is a bit simpler than setting up a separate endpoint.

You can explicitly pass values _from_ the JSON file that Google provides to the auth client.  Your setup will look something like this (with a service account + Google Analytics 4 API as an example, but replace with any Google API):

```ts
import { GoogleAuth } from 'google-auth-library' // no need to install this library, it comes with @google-analytics/data
import { BetaAnalyticsDataClient } from '@google-analytics/data'

const cli = new BetaAnalyticsDataClient({
  auth: new GoogleAuth({
      projectId: 'yourprojectid',
      scopes: 'https://www.googleapis.com/auth/analytics',
      credentials: {
        client_email: 'email from json file',
        private_key: 'pk from json file'
      })
})

// Now you can use the relevant API
cli.runReport({ ... }).then()
```

**Reply by UnlockingLanguage** [reply]
This was a lifesaver btw 10/10 @zachgoll

**Reply by caweidmann** [reply]
Thank you! Works like a charm.

**Comment by nuotsu** [comment]
Looks like you can include files with `config.experimental.outputFileTracingIncludes` per the [official docs here](https://nextjs.org/docs/pages/api-reference/next-config-js/output#caveats), replacing `'api/data'` with whatever route uses the file:

```js
// next.config.js
const nextConfig = {
  experimental: {
    outputFileTracingIncludes: {
      'api/data': ['./secrets.json'],
    },
  },
}
```
